crates/re_data_ui/src/image.rs

```diff
@@ -230,26 +232,77 @@ fn texture_size(colormapped_texture: &ColormappedTexture) -> Vec2 {
     egui::vec2(w as f32, h as f32)
 }
 
-fn show_image_at_max_size(
+#[derive(Clone, Copy, PartialEq, Eq)]
+enum PreviewSizeShrink {
+    /// Stick to the desired size. May cause
+    DontShrink,
+
+    /// The allocated size may shrink in one dimension to fit the aspect ratio of the image.
+    ShrinkToTextureAspectRatio,
+}
+
+/// Shows preview of an image.
+///
+/// Displays the image in inside the desired size.
+/// Image will be stretched in the dimensions it is below a certain threshold.
+///
+/// TODO(andreas): Small images will be blown up to the desired size, this may not always be what we want.
+///                 Maybe add `PreviewSizeShrink::ShrinkToFitWithMin(egui::Vec2)`?
+fn show_image_preview(
     render_ctx: &mut re_renderer::RenderContext,
     re_ui: &ReUi,
     ui: &mut egui::Ui,
     colormapped_texture: ColormappedTexture,
     debug_name: &str,
-    max_size: Vec2,
+    desired_size: egui::Vec2,
+    image_fit: PreviewSizeShrink,
 ) -> egui::Response {
-    let desired_size = {
-        let mut desired_size = texture_size(&colormapped_texture);
-        desired_size *= (max_size.x / desired_size.x).min(1.0);
-        desired_size *= (max_size.y / desired_size.y).min(1.0);
-        desired_size
+    let texture_size = texture_size(&colormapped_texture);
+
+    // First we figure out how much we should scale (uniformly) the image.
+    let texture_scale_factor = {
+        // Make the image at least this thick both x and y (unless desired_size explicitly asks for smaller)
+        const MIN_SIZE: f32 = 2.0;
+        let min_size = egui::Vec2::splat(MIN_SIZE).at_most(desired_size);
+
+        // Ideally, we just scale down/up to fit into the available space.
+        let zoom = (desired_size / texture_size).min_elem();
+
+        // But it can happen that this makes one of the axis too small, so let's stretch it if needed.
+        egui::Vec2::splat(zoom).at_least(min_size / texture_size)
+    };
+
+    let scaled_texture_size = texture_size * texture_scale_factor;
+
+    // Request the desired size, but shrink one dimension if the texture is too wide/narrow if so required.
+    let requested_rect = match image_fit {
+        PreviewSizeShrink::DontShrink => desired_size,
+        PreviewSizeShrink::ShrinkToTextureAspectRatio => {
```

Comment:
I don't follow this logic - we should always respect aspect ratio, with the single exception of very thin images - and that exception is always a valid exception, afaict. So why is there an option here to do different things? Shouldn't we always resize the image to the given desired_size (in an aspect-ratio-respecting manner) and then slightly adjust one axis if it becomes thinner than `MIN_SIZE`?

Reply:
Maybe it's just poorly named: What I'm after is that the reserved ui space would adapt to the texture or not.

Situation where we'd like the ui element for the preview to take a fixed amount of space:
<img width="287" alt="image" src="https://github.com/rerun-io/rerun/assets/1220815/08ff7617-374e-4906-9f4d-d83cb2ae4bb6">
-> If this icon is a vertical gradient, I still want it to take the same amount of space left and right because otherwise it doesn't align well if there's more such lines below, causing a less tidy look

This here on the other hand is pretty much the only situation where I do want the ui space to adapt to the aspect ratio of the image.
<img width="472" alt="image" src="https://github.com/rerun-io/rerun/assets/1220815/8575eae2-002d-4806-bf0b-685313f47a3b">
Having always the 400x400 box here would be pretty weird for a thin gradient, we want to shrink the ui around it without making the image smaller.

Reply:
So this option isn't really about what happens with the texture but what happens with the allocated space. Naming should reflect this better.

But before I touch stuff again, do you agree that the outlined behavior makes sense or do you think this is weird/confusing/wrong?

Reply:
> If this icon is a vertical gradient, I still want it to take the same amount of space left and right because otherwise it doesn't align well if there's more such lines below, causing a less tidy look

That makes sense - in the table you want all preview images to take up the same amount of space.

But that can be made a lot simpler with:

```rs
ui.allocate_ui(max_image_size, |ui| {
    show_image_preview(…, max_image_size)
});
```

no need for `show_image_preview` to worry about all that complexity.

(also, `show_image_preview` should be called `image_preview_ui`)